### ISSUE TITLE
Extract core features to independent sub-crates

### ISSUE THREAD

**Issue by ubnt-intrepid** [websockets-rs/rust-websocket#180]
Currently, `rust-websocket` provides the all features from a single crate, but in the case of using only certain features (like encoding/decoding data frames) most of these features are unneeded.

It is desirable to extract some components which do not depend on the higher level implementations (such as client and server) as independent crates.

**Comment by vi** [comment]
It's a good idea.

What architecture (and which item goes to which crate) do you suggest? Shall it be just two crates: IO-agnostic (without references to tokio or `std::net`) and high-level? Or something more elaborate?

The next major-ish version would probably be just updating websocket-rs's dependencies to recent versions.

**Comment by illegalprime** [comment]
I agree, it sounds really good. Maybe removing the `hyper` dependency is a good motivator to split this up into many crates, its already rewriting a lot of it anyway.
